ERCS/erc-7683.md

```diff
@@ -24,26 +24,47 @@ By implementing a standard, cross-chain intents systems can interoperate and sha
 
 The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
 
-### CrossChainOrder struct
+### Glossary of Terms
 
-A compliant cross-chain order type MUST be ABI decodable into the `CrossChainOrder` type.
+- **Destination Chain**: the chain where the intent is executed and the user receives funds. Note: intents can involve multiple destination chains.
+- **Filler**: a participant who fulfils a user intent on the destination chain(s) and receives payment as a reward.
+- **Leg**: a portion of the user intent that can be executed independently from others. All legs must be executed for an intent to be considered fulfilled.
+- **Origin chain**: the chain where the user sends funds.
+- **Settlement System**: a system that custodies user deposits, verifies fills, and pays fillers for the purpose of faciliating intents.
+- **Settler**: a contract that implements part of the settlement system on a particular chain.
+- **User**: for the purposes of this document, the user is the end-user who is sending the ERC-7683 order.
+
+### Order structs
+
+A compliant cross-chain order type MUST be ABI decodable into either `GaslessCrossChainOrder` or `OnchainCrossChainOrder` type.
 
 ```solidity
-/// @title CrossChainOrder type
-/// @notice Standard order struct to be signed by swappers, disseminated to fillers, and submitted to settlement contracts
-struct CrossChainOrder {
+/// @title GaslessCrossChainOrder CrossChainOrder type
+/// @notice Standard order struct to be signed by users, disseminated to fillers, and submitted to origin settler contracts
+struct GaslessCrossChainOrder {
 	/// @dev The contract address that the order is meant to be settled by.
 	/// Fillers send this order to this contract address on the origin chain
-	address settlementContract;
+	address originSettler;
 	/// @dev The address of the user who is initiating the swap,
 	/// whose input tokens will be taken and escrowed
-	address swapper;
+	address user;
 	/// @dev Nonce to be used as replay protection for the order
 	uint256 nonce;
```

Comment:
If the Gasless Order does not contain an explicit orderId, then solvers won't be able to pre fill the order on the destination chain before calling open on the source chain because they won't be able to deterministically compute the orderId. 

Is there a reason why this nonce should differ from the orderId emitted in the Open event. Maybe we could reuse this as the orderId, instead of creating a new field for it.

Reply:
Does the standard imply that the nonce cannot be the orderId? I don't see why you couldn't do this.

If I were designing a pre-fill system, I might not use the naked nonce as the orderId (depending on the way the offchain order system is designed). Instead, I would probably hash the nonce with other fields that can't be spoofed, like user _and/or_ msg.sender, so I can be sure the nonce can't be stolen by a malicious 3rd party that frontruns me.

Reply:
Yeah makes sense. But this requires the fillers to be aware of the logic that the settlement system will use to generate the orderId, and if they accidentally get this logic wrong, then they lose money, because they fill for an orderId that can't exist.

But I understand why it wouldn't make sense to add an explicitly orderId into the struct , because it would take away the ability for settlement systems to maintain an ordered orderId on chain.

I think for our case we can just an orderId field to our orderData, which is constructed in a determinstic way like you mentioned.

Reply:
> But this requires the fillers to be aware of the logic that the settlement system will use to generate the orderId, and if they accidentally get this logic wrong, then they lose money, because they fill for an orderId that can't exist.

This is a good point. I think this was an oversight. Predicting the `orderId` should be part of the standard.

My intention was that fillers could use resolve to predict their orderId, but now I realize that they can't do that because `orderId` isn't included in the struct returned by resolve. The two alternatives I see are:
- New method where you pass in an order and get an `orderId`. Problem: fillers need to call two methods if trying to determine their fill arguments.
- Add `orderId` to `ResolvedCrossChainOrder`. Problem: it creates redundancy in the `Open` event (`orderId` is emitted in the event in two places), which is inefficient.